### PR TITLE
Add async version of send_message for both tcp and rtu

### PR DESCRIPTION
**Pull request by tiagocoutinho**
Now that #86 is out I feel more comfortable making this PR since the new async/await python 3 syntax makes it impossible to use with python 2.

This PR proposes to add an async version of send_message for both tcp and rtu.
We will be able to write asyncio based clients like this:

```python
import asyncio

from umodbus import conf
from umodbus.client import tcp

async def main():
    reader, writer = await asyncio.open_connection('localhost', 502)
    message = tcp.write_multiple_coils(slave_id=1, starting_address=1, values=[1, 0, 1, 1])
    response = await tcp.async_send_message(message, reader, writer)
    writer.close()
    await writer.wait_closed()


asyncio.run(main())
```

We are not restricted to TCP. Any I/O library supporting [asyncio](https://docs.python.org/3/library/asyncio.html) [StreamReader](https://docs.python.org/3/library/asyncio-stream.html#streamreader) and [StreamWriter](https://docs.python.org/3/library/asyncio-stream.html#streamwriter) interface (example: Using [serial_asyncio](https://github.com/pyserial/pyserial-asyncio)) can be used with either the TCP or RTU versions of `async_send_message`.

### CODE REVIEW

**Comment by OrangeTux** [comment]
Thanks for this PR! I might need a bit of time to review the PR and test it myself. I'll come back to you.

**Comment by tiagocoutinho** [comment]
You're welcome.
No worries, take your time.

File: README.rst [hunk 1]
```diff
@@ -103,6 +103,39 @@ Doing a Modbus request requires even less code:
...
+
+    from umodbus import conf
+    from umodbus.client import tcp
+
+    # Enable values to be signed (default is False).
+    conf.SIGNED_VALUES = True
+
+
+    async def main():
+        reader, writer = await asyncio.open_connection('localhost', 15020)
```
Comment (jaapz) [inline README.rst]: Nitpick: would be nice to use the same 502 port as in the example above

File: scripts/examples/simple_async_tcp_client.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python
+# scripts/examples/simple_async_tcp_client.py
```
Comment (jaapz) [inline scripts/examples/simple_async_tcp_client.py]: Nitpick: name of this file should probably be `simple_asyncio_tcp_client.py` (in stead of async)

File: tests/system/responses/test_exception_async_responses.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    (3, 0),
+    (4, 0),
+    (1, 0x07D0 + 1),
+    (2, 0x07D0 + 1),
+    (3, 0x007D + 1),
+    (4, 0x007D + 1),
+])
+async def test_request_returning_invalid_data_value_error(async_tcp_streams, mbap, function_code,
+                                                    quantity):
+    """ Validate response PDU of request returning excepetion response with
```
Comment (jaapz) [inline tests/system/responses/test_exception_async_responses.py]:
```suggestion
    """ Validate response PDU of request returning exception response with
```

File: tests/system/responses/test_exception_async_responses.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    (2, 0),
+    (3, 0),
+    (4, 0),
+    (1, 0x07D0 + 1),
+    (2, 0x07D0 + 1),
+    (3, 0x007D + 1),
+    (4, 0x007D + 1),
+])
+async def test_request_returning_invalid_data_value_error(async_tcp_streams, mbap, function_code,
+                                                    quantity):
```
Comment (jaapz) [inline tests/system/responses/test_exception_async_responses.py]: Nitpick: Indent of this quantity argument does not line up with the rest of the argument

File: tests/system/responses/test_exception_async_responses.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    """ Validate response PDU of request returning excepetion response with
+    error code 3.
+    """
+    function_code, starting_address, quantity = (function_code, 0, quantity)
```
Comment (jaapz) [inline tests/system/responses/test_exception_async_responses.py]: Why do you pack and then unpack these variables, couldn't you just do

```suggestion
    starting_address = 0
```

here?

File: tests/system/responses/test_exception_async_responses.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    (partial(tcp.read_discrete_inputs, 1, 9, 2)),
+    (partial(tcp.read_holding_registers, 1, 9, 2)),
+    (partial(tcp.read_input_registers, 1, 9, 2)),
+    (partial(tcp.write_single_coil, 1, 11, 0)),
+    (partial(tcp.write_single_register, 1, 11, 1337)),
+    (partial(tcp.write_multiple_coils, 1, 9, [1, 1])),
+    (partial(tcp.write_multiple_registers, 1, 9, [1337, 15])),
+])
+async def test_request_returning_invalid_data_address_error(async_tcp_streams, function):
+    """ Validate response PDU of request returning excepetion response with
```
Comment (jaapz) [inline tests/system/responses/test_exception_async_responses.py]:
```suggestion
    """ Validate response PDU of request returning exception response with
```

File: tests/system/responses/test_exception_async_rtu_responses.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    (3, 0),
+    (4, 0),
+    (1, 0x07D0 + 1),
+    (2, 0x07D0 + 1),
+    (3, 0x007D + 1),
+    (4, 0x007D + 1),
+])
+async def test_request_returning_invalid_data_value_error(
+    rtu_server, async_serial_streams, function_code, quantity):
+    """ Validate response PDU of request returning excepetion response with
```
Comment (jaapz) [inline tests/system/responses/test_exception_async_rtu_responses.py]:
```suggestion
    """ Validate response PDU of request returning exception response with
```

File: tests/system/responses/test_exception_async_responses.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    mbap = adu[:7]
+    function_code = struct.unpack('>B', adu[7:8])[0]
+
+    reader, writer = async_tcp_streams
+    writer.write(adu)
+    await writer.drain()
+    resp = await reader.read(1024)
+
+    validate_response_mbap(mbap, resp)
+    assert struct.unpack('>BB', resp[-2:]) == (0x80 + function_code, 4)
```
Comment (jaapz) [inline tests/system/responses/test_exception_async_responses.py]: Lines 86-92 seem to be recurring in all these tests, maybe you could factor them out into a utility or a fixture and clean up the tests some more, you could call it something like `assert_function_code_in_response(...)` or something.

File: tests/system/responses/test_exception_async_rtu_responses.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    adu = add_crc(struct.pack('>BBHH', slave_id, function_code,
+                              starting_address, quantity))
+
+    reader, writer = async_serial_streams
+    writer.write(adu)
+    await writer.drain()
+    resp = await reader.read(rtu_server.serial_port.in_waiting)
+
+    validate_crc(resp)
+    assert struct.unpack('>BB', resp[1:-2]) == (0x80 + function_code, 3)
```
Comment (jaapz) [inline tests/system/responses/test_exception_async_rtu_responses.py]: Referring back to a previous comment, line 35-41 are pretty similar to those in other tests, could you factor them  out into a utility?

File: tests/system/responses/test_exception_async_rtu_responses.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    (partial(rtu.read_discrete_inputs, 1, 9, 2)),
+    (partial(rtu.read_holding_registers, 1, 9, 2)),
+    (partial(rtu.read_input_registers, 1, 9, 2)),
+    (partial(rtu.write_single_coil, 1, 11, 0)),
+    (partial(rtu.write_single_register, 1, 11, 1337)),
+    (partial(rtu.write_multiple_coils, 1, 9, [1, 1])),
+    (partial(rtu.write_multiple_registers, 1, 9, [1337, 15])),
+])
+async def test_request_returning_invalid_data_address_error(rtu_server, async_serial_streams, function):
+    """ Validate response PDU of request returning excepetion response with
```
Comment (jaapz) [inline tests/system/responses/test_exception_async_rtu_responses.py]:
```suggestion
    """ Validate response PDU of request returning exception response with
```

File: tests/system/responses/test_exception_async_rtu_responses.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    (partial(rtu.read_discrete_inputs, 1, 666, 1)),
+    (partial(rtu.read_holding_registers, 1, 666, 1)),
+    (partial(rtu.read_input_registers, 1, 666, 1)),
+    (partial(rtu.write_single_coil, 1, 666, 0)),
+    (partial(rtu.write_single_register, 1, 666, 1337)),
+    (partial(rtu.write_multiple_coils, 1, 666, [1])),
+    (partial(rtu.write_multiple_registers, 1, 666, [1337])),
+])
+async def test_request_returning_server_device_failure_error(rtu_server, async_serial_streams, function):
+    """ Validate response PDU of request returning excepetion response with
```
Comment (jaapz) [inline tests/system/responses/test_exception_async_rtu_responses.py]:
```suggestion
    """ Validate response PDU of request returning exception response with
```

File: tests/system/responses/test_succesful_async_responses.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+@pytest.mark.parametrize('function', [
+    tcp.read_coils,
+    tcp.read_discrete_inputs,
+])
+async def test_response_on_single_bit_value_read_requests(async_tcp_streams, function):
+    """ Validate response of a succesful Read Coils or Read Discrete Inputs
+    request.
+    """
+    slave_id, starting_address, quantity = (1, 0, 10)
```
Comment (jaapz) [inline tests/system/responses/test_succesful_async_responses.py]: Why not just

```suggestion
    slave_id = 1
    starting_address = 0
    quantity = 10
```

I mean yeah unpacking a set saves you two lines but IMHO it's just slightly less readable

Comment (jaapz) [reply tests/system/responses/test_succesful_async_responses.py]: This comment applies to every instance of this in your PR

File: umodbus/client/serial/rtu.py [hunk 13]
```diff
@@ -223,3 +223,27 @@ def send_message(adu, serial_port):
         serial_port.read, expected_response_size - exception_adu_size)
 
     return parse_response_adu(response_error_adu + response_remainder, adu)
+
+
+async def async_send_message(adu, reader, writer):
```
Comment (jaapz) [inline umodbus/client/serial/rtu.py]: Would be nice to also have an example file using this function. It isn't completely clear to me how I would use this with an actual serial device, as I would with the non-asyncio `send_message` function.

**Review by jaapz — CHANGES_REQUESTED** [review]
Hey @tiagocoutinho , thanks for your contribution. This is a great addition to uModbus. Just had some comments for you to look at, before we can merge.

I'm personally also not a big fan of the `async_send_message` names. Maybe the async method could be moved to an `asyncio` module so it could be used like

```python
from umodbus.client.tcp.asyncio import send_message
await send_message(...)
```

File: README.rst (hunk 1 above)
Comment (tiagocoutinho) [reply README.rst]: Makes total sense. Nice catch!

File: scripts/examples/simple_async_tcp_client.py (hunk 2 above)
Comment (tiagocoutinho) [reply scripts/examples/simple_async_tcp_client.py]: Ok.
Have in mind that although I called the branch *asyncio-client*, in fact, the entire PR is async library agnostic.
But, of course, the example uses *asyncio* explicitly so it makes total sense to change the name.

File: tests/system/responses/test_succesful_async_responses.py (hunk 12 above)
Comment (tiagocoutinho) [reply tests/system/responses/test_succesful_async_responses.py]: I was just trying to be consistent with *test_succesful_responses.py*

File: umodbus/client/serial/rtu.py (hunk 13 above)
Comment (tiagocoutinho) [reply umodbus/client/serial/rtu.py]: Sure, I just added one: `scirpts/examples/simple_asyncio_rtu_client.py`.

File: tests/system/responses/test_exception_async_responses.py (hunk 4 above)
Comment (tiagocoutinho) [reply tests/system/responses/test_exception_async_responses.py]: Of course. Done!

File: tests/system/responses/test_exception_async_responses.py (hunk 5 above)
Comment (tiagocoutinho) [reply tests/system/responses/test_exception_async_responses.py]: Of course. Done!

File: tests/system/responses/test_exception_async_rtu_responses.py (hunk 9 above)
Comment (tiagocoutinho) [reply tests/system/responses/test_exception_async_rtu_responses.py]: Sure, it should be better now.

File: tests/system/responses/test_succesful_async_responses.py (hunk 12 above)
Comment (tiagocoutinho) [reply tests/system/responses/test_succesful_async_responses.py]: It should be better now.

**Comment by tiagocoutinho** [comment]
Hi @jaapz,
Thanks for your review.

I am in favour of moving the `async_send_message()` to it's own module and renaming the function to `send_message()`.
As you probably noticed, I there is no `import asyncio` in the PR. This is intentional and allows umodbus to state that is is *async library agnostic*.
The idea was that it would be simple to use umodbus in an [asyncio](https://docs.python.org/3/library/asyncio.html) application and, by writing a simple adapter, possible to use in a [curio](https://curio.rtfd.io), [trio](https://trio.rtfd.io) or [anyio](https://anyio.rtfd.io) application.

So, long story short, what do you thing about naming the module `umodbus.client.tcp.asynch` instead?

**Comment by jaapz** [comment]
@tiagocoutinho that makes complete sense, calling it `umodbus.client.tcp.async` instead is fine by me. Maybe you could also add a bit of documentation to the readme stating that we are in fact async library agnostic?

**Comment by tiagocoutinho** [comment]
> Maybe you could also add a bit of documentation to the readme stating that we are in fact async library agnostic?

You read my mind :-)

**Comment by tiagocoutinho** [comment]
I think it should be ready for review now.
I re-based from master in order to update the examples with the argument parser.
I took the liberty of adding also a reference to gevent in the README and I added a curio tcp client example.

File: README.rst [hunk 14]
```diff
@@ -126,6 +128,62 @@ Doing a Modbus request requires even less code:
...
+As an example, the above client snippet can be made to work in a gevent
+context simply by replacing the ``import socket`` line with
+``from gevent import socket``.
+
+Here is the same request using any asyncio_ compatible StreamReader and
+StreamWriter objects:
+
+..
+    Because GitHub doesn't support the include directive the source of
+    scripts/examples/simple_tcp_client.py has been copied to this file.
```
Comment (jaapz) [inline README.rst]:
```suggestion
    scripts/examples/simple_async_tcp_client.py has been copied to this file.
```

copy paste error :)

File: umodbus/client/tcp/__init__.py [hunk 15]
```diff
@@ -0,0 +1 @@
+from .tcp import *
```
Comment (jaapz) [inline umodbus/client/tcp/__init__.py]: why not move the `tcp.py` file contents to this `__init__.py`? Saves an import and an extra file

File: tests/unit/test_tcp.py [hunk 16]
```diff
@@ -1,6 +1,6 @@
 import struct
 
-from umodbus.client.tcp import _create_request_adu, _create_mbap_header
+from umodbus.client.tcp.tcp import _create_request_adu, _create_mbap_header
```
Comment (jaapz) [inline tests/unit/test_tcp.py]: Is this necessary even though there is a * import in the umodbus.client.tcp.__init__?

File: umodbus/client/tcp/asynch.py [hunk 17]
```diff
@@ -0,0 +1,29 @@
+"""
```
Comment (jaapz) [inline umodbus/client/tcp/asynch.py:1]: I think the filename is `asynch.py` as opposed to `async.py` because `async` is a protected keyword, right? Would be nice if we could call it `async` but I guess if Python doesn't let us, this is the next best thing.

**Review by jaapz — CHANGES_REQUESTED** [review]
Just a few small changes/comments, almost there! Again, great work, thanks for your contributions and thanks for bearing with me during the review.

**Comment by jaapz** [comment]
I've done another small review, everything else looks good to me. Maybe @OrangeTux can take over from here to test it a bit and get this merged and released?

File: umodbus/client/tcp/__init__.py (hunk 15 above)
Comment (tiagocoutinho) [reply umodbus/client/tcp/__init__.py]: I was tempted to do it but I noticed in the project all `__init__.py` files are empty so I aimed for consistency.
I move the contents of `tcp.py` then to `__init__.py`

File: tests/unit/test_tcp.py (hunk 16 above)
Comment (tiagocoutinho) [reply tests/unit/test_tcp.py]: Yes, the "*" only imports *visible* symbols. See [here](https://s16h.medium.com/importing-star-in-python-88fe9e8bd4d2) for a better explanation.
I was also surprised. I guess it is a python 3 feature. I don't remember python 2 behaving like this.

If I move everything to `__init__.py` this won't be an issue anymore. So +1 for moving the code to `__init__.py` :-)

File: umodbus/client/tcp/asynch.py (hunk 17 above)
Comment (tiagocoutinho) [reply umodbus/client/tcp/asynch.py:1]: Yes, exactly.

**Comment by tiagocoutinho** [comment]
Hi @OrangeTux , @jaapz,
Just pinging... Does everything look ok with the PR? 
I don't see any pending changes. Please let me know if you are waiting on me.
Thanks

**Comment by jaapz** [comment]
Hey @tiagocoutinho, I just need some time to properly test this before merging. Sorry for the wait, I'm pretty busy at the moment
